File: sol/2316.py
class Solution:
    def countPairs(self, n: int, edges: List[List[int]]) -> int:
        parent = list(range(n))
        size = [1] * n
        def getParent(x):
            while parent[x] != x:
                x, parent[x] = parent[x], parent[parent[x]]
            return x
        def union(x, y):
            x = getParent(x)
            y = getParent(y)
            if x == y:
                return
            if size[x] < size[y]:
                x, y = y, x
            parent[y] = x
            size[x] += size[y]
        for i, j in edges:
            union(i, j)
        res = 0
        prefix = 0
        for c in (size[x] for x in frozenset(getParent(i) for i in range(n))):
            res += prefix * c
            prefix += c
        return res

File: sol/2316/2316.cpp
#include <unordered_set>

struct Solution {
    int getPar(int x, vector<int> &par){
        while (x != par[x]){
            int tmp = x;
            x = par[tmp];
            par[tmp] = par[par[tmp]];
        }
        return x;
    }
    inline void unionGp(int x, int y, vector<int> &par, vector<int> &size){
        x = this->getPar(x, par);
        y = this->getPar(y, par);
        if (x == y){
            return;
        }
        if (size[x] < size[y]){
            swap(x, y);
        }
        par[y] = x;
        size[x] += size[y];
    }
    long long countPairs(int n, vector<vector<int>>& edges) {
        vector<int> par(n, 0), size(n, 1);
        iota(par.begin(), par.end(), 0);
        for (auto &e : edges){
            this->unionGp(e[0], e[1], par, size);
        }
        unordered_set<int> heads;
        for (int i = 0; i < n; ++i){
            heads.insert(this->getPar(i, par));
        }
        long long res = 0, prefix = 0;
        for (int i : heads){
            int x = size[i];
            res += x * prefix;
            prefix += x;
        }
        return res;
    }
};
